core/priority/component.go

```diff
@@ -0,0 +1,292 @@
+	Score    int
+}
+
+// coreConsensus is an interface for the core/consensus.Component.
+type coreConsensus interface {
+	ProposePriority(context.Context, core.Duty, *pbv1.PriorityResult) error
+	SubscribePriority(func(context.Context, core.Duty, *pbv1.PriorityResult) error)
+}
+
+// NewComponent returns a new priority component.
+func NewComponent(tcpNode host.Host, peers []peer.ID, minRequired int, sendFunc p2p.SendReceiveFunc,
+	registerHandlerFunc p2p.RegisterHandlerFunc, consensus coreConsensus,
+	slotDuration time.Duration, privkey *ecdsa.PrivateKey,
+) (*Component, error) {
+	verifier, err := newMsgVerifier(peers)
+	if err != nil {
+		return nil, err
+	}
+
+	tickerProvider := func() (<-chan time.Time, func()) {
+		ticker := time.NewTicker(slotDuration / 10)
+		return ticker.C, ticker.Stop
+	}
+
+	prioritiser := newInternal(tcpNode, peers, minRequired, sendFunc, registerHandlerFunc,
+		consensusWrapper{consensus}, verifier, slotDuration, tickerProvider)
+
+	return &Component{
+		prioritiser: prioritiser,
+		privkey:     privkey,
+	}, nil
+}
+
+// Component wraps a prioritise protocol instance providing a friendly API (hiding the underlying protobuf types) and does signing.
+type Component struct {
+	peerID      peer.ID
+	privkey     *ecdsa.PrivateKey
+	prioritiser *Prioritiser
+}
+
+// Subscribe registers a prioritiser output subscriber function.
+func (c *Component) Subscribe(fn func(context.Context, core.Duty, []TopicResult) error) {
+	c.prioritiser.Subscribe(func(ctx context.Context, instance Instance, result *pbv1.PriorityResult) error {
+		dutypb, ok := instance.(*pbv1.Duty)
+		if !ok {
+			return errors.New("invalid duty instance")
+		}
+
+		var results []TopicResult
+		for _, topic := range result.Topics {
+			result, err := topicResultFromProto(topic)
+			if err != nil {
+				return err
+			}
+
+			results = append(results, result)
+		}
+
+		return fn(ctx, core.DutyFromProto(dutypb), results)
+	})
+}
+
+// Prioritise starts a new prioritisation instance for the provided duty and proposals or returns an error.
+func (c *Component) Prioritise(ctx context.Context, duty core.Duty, proposals ...TopicProposal) error {
+	instance, err := anypb.New(core.DutyToProto(duty))
+	if err != nil {
+		return errors.Wrap(err, "any proto duty")
+	}
+
+	var topics []*pbv1.PriorityTopicProposal
+	for _, proposal := range proposals {
+		proposalPB, err := topicProposalToProto(proposal)
+		if err != nil {
+			return err
+		}
+
+		topics = append(topics, proposalPB)
+	}
+
+	msg := &pbv1.PriorityMsg{
+		Instance: instance,
+		PeerId:   c.peerID.String(),
+		Topics:   topics,
+	}
+
+	msg, err = signMsg(msg, c.privkey)
+	if err != nil {
+		return err
+	}
+
+	return c.prioritiser.Prioritise(ctx, msg)
+}
+
+// signMsg returns a copy of the proto message with a populated signature signed by the provided private key.
+func signMsg(msg *pbv1.PriorityMsg, privkey *ecdsa.PrivateKey) (*pbv1.PriorityMsg, error) {
+	clone := proto.Clone(msg).(*pbv1.PriorityMsg)
```

Comment:
what if msg is nil? `clone.Signature` would panic

Reply:
nil checks are not required in golang, one assumes valid data is being passed in 99% of cases